builtin/provisioners/file/resource_provisioner.go

```diff
@@ -48,13 +43,52 @@ func (p *ResourceProvisioner) Apply(
 
 // Validate checks if the required arguments are configured
 func (p *ResourceProvisioner) Validate(c *terraform.ResourceConfig) (ws []string, es []error) {
-	v := &config.Validator{
-		Required: []string{
-			"source",
-			"destination",
-		},
+	numDst := 0
+	numSrc := 0
+	for name := range c.Raw {
+		switch name {
+		case "destination":
+			numDst++
+		case "source", "content":
+			numSrc++
+		default:
+			es = append(es, fmt.Errorf("Unknown configuration '%s'", name))
+		}
+	}
+	if numSrc != 1 || numDst != 1 {
+		es = append(es, fmt.Errorf("Must provide one  of 'content' or 'source' and 'destination' to file"))
+	}
+	return
+}
+
+// getSrc returns the file to use as source
+func (p *ResourceProvisioner) getSrc(c *terraform.ResourceConfig) (string, error) {
+	var src string
+
+	sRaw, ok := c.Config["source"]
+	if ok {
+		if src, ok = sRaw.(string); !ok {
+			return "", fmt.Errorf("Unsupported 'source' type! Must be string.")
+		}
+	}
+
+	content, ok := c.Config["content"]
+	if ok {
+		file, err := ioutil.TempFile("", "tf-file-content")
```

Comment:
Just a thought, why use a TempFile? Using a tempfile means that the string has to be written to disk, then read up again in the copyFiles function and then the file should be deleted, from the documentation "It is the caller's responsibility to remove the file when no longer needed". Since the function Communicator.Upload takes a io.Reader so this could be handled with a byte buffer, so no writes to disk, and no orphaned tempfile gets left behind. Created a gist with this implementation in mind. [https://gist.github.com/HX-Rd/80eca9df30da1eddc61ab0e077452e55](url)